Separate ca_names handling for client and server

mattcaswell opened the pull request:
SSL(_CTX)?_set_client_CA_list() was a server side only function in 1.1.0.
If it was called on the client side then it was ignored. In 1.1.1 it now
makes sense to have a CA list defined for both client and server (the
client now sends it the the TLSv1.3 certificate_authorities extension).
Unfortunately some applications were using the same SSL_CTX for both
clients and servers and this resulted in some client ClientHellos being
excessively large due to the number of certificate authorities being sent.

This commit seperates out the CA list updated by
SSL(_CTX)?_set_client_CA_list() and the more generic
SSL(_CTX)?_set0_CA_list(). This means that SSL(_CTX)?_set_client_CA_list()
still has no effect on the client side. If both CA lists are set then
SSL(_CTX)?_set_client_CA_list() takes priority.

Fixes #7411

mattcaswell commented:
Ping @vdukhovni.

vdukhovni commented:
Matt, how do you want to proceed?

vdukhovni commented:
> @mattcaswell I'd like to proceed by searching for enough cycles to properly respond to your comments and see what I can do with the documentation :-)

Touché!  Fair enough. :-)

mattcaswell commented:
I rebased this to resolve a merge conflict. All feedback has been addressed. I merged together 3 different man pages and rewrote some of the text to try and address the concerns.

mattcaswell commented:
Ping @vdukhovni

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -2,9 +2,20 @@
+SSL_CTX_add_client_CA,
+SSL_add_client_CA,
+SSL_get0_peer_CA_list
+- get or set CA list
```
Thanks for merging the manpages I think this is better.

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -17,10 +28,27 @@ SSL_get0_peer_CA_list - get or set CA list
 
+The following functions can be called on either the server or the client side.
+For TLS versions before 1.3 the list of CA names is only sent from the server
+to client when requesting a client certificate. So any list of CA names set
```
s/from the server to client/from the server to the client/ or else "from server to client".

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -17,10 +28,27 @@ SSL_get0_peer_CA_list - get or set CA list
 =head1 DESCRIPTION
 
+The following functions can be called on either the server or the client side.
+For TLS versions before 1.3 the list of CA names is only sent from the server
```
No strong views, would "1.2 and earlier" be more clear than "before 1.3" or not?

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -17,10 +28,27 @@ SSL_get0_peer_CA_list - get or set CA list
 
 =head1 DESCRIPTION
 
+The following functions can be called on either the server or the client side.
```
I think more clear to say `the below functions` rather than `the following functions`, and list them as soon as possible, so that the context is clear.

vdukhovni replied:
I would also, as early as possible, discourage the use of the new functions to set a non-empty CA list in any client or dual-purpose client+server context.  The sending of client CA lists to the server is generally unnecessary, is a privacy leak and risks interoperability issues when the client HELLO gets too large.  Users should generally stick the older `_client_` functions, and perhaps those should even be described first?

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -17,10 +28,27 @@ SSL_get0_peer_CA_list - get or set CA list
+SSL_get0_peer_CA_list() is always B<NULL>.
+
+For TLS 1.3 the list of CA names is sent using the B<certificate_authorities>
+extension and will be sent by a client (in the ClientHello message) or by
```
s/will be/may be/   (since the client extension is not sent if empty IIRC).

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -42,47 +71,124 @@ SSL_add1_to_CA_list() appends the CA subject name extracted from B<x> to the
 
-SSL_get0_peer_CA_list() retrieves the list of CA names (if any) the peer
-has sent.
+The following functions are very similar to the functions described above except
```
"Below" vs. "Following"?

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -42,47 +71,124 @@ SSL_add1_to_CA_list() appends the CA subject name extracted from B<x> to the
-has sent.
+The following functions are very similar to the functions described above except
+that they only have an effect when called on the server side. The list of CA
+names managed are separate - so you cannot (for example) set CA names using the
```
s/the list of/the lists of/

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -42,47 +71,124 @@ SSL_add1_to_CA_list() appends the CA subject name extracted from B<x> to the
+names managed are separate - so you cannot (for example) set CA names using the
+functions above and then get them using the functions below. Where a mix of the
+above functions and the functions below have been used on the server side the
+settings set by the following functions take precedence. These functions are
```
Here, instead of "the above" and the "below" I would suggest some more direct binding, such as "dual-purpose" and "client-specific", or similar.  Indeed perhaps name to the two classes of functions when first introducing them, so that the category names are already "established" by the time we get to this point.

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -42,47 +71,124 @@ SSL_add1_to_CA_list() appends the CA subject name extracted from B<x> to the
+above functions and the functions below have been used on the server side the
+settings set by the following functions take precedence. These functions are
+present for backwards compatibility. L<SSL_CTX_set0_CA_list(3)> and similar
+functions should be used in preference.
```
Here, I strongly disagree. The new functions are __only__ for the few and the brave who want to signal CA names from client to server, which should almost never happen, and they should have been duly warned.  The new functions expose a new TLS 1.3 feature, but it is one that should generally be avoided.  The preferred, safer, interface is provided by the older functions.

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -42,47 +71,124 @@ SSL_add1_to_CA_list() appends the CA subject name extracted from B<x> to the
+if the list is large. This optional capability was introduced as part of TLSv1.3
+and therefore setting CA names on the client side will have no impact if that
+protocol version has been disabled. Most servers do not need this and so this
+should be avoided unless required.
```
The warning should IMHO happen early, as part of introducing the new API, after covering the original preferred API.

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -42,47 +71,124 @@ SSL_add1_to_CA_list() appends the CA subject name extracted from B<x> to the
+L<SSL_CTX_load_verify_locations(3)> to additionally load them for verification.
+
+If the list of acceptable CAs is compiled in a file, the
+L<SSL_load_client_CA_file(3)> function can be used to help importing the
```
s/importing/to import/

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -42,47 +71,124 @@ SSL_add1_to_CA_list() appends the CA subject name extracted from B<x> to the
+Scan all certificates in B<CAfile> and list them as acceptable CAs:
+
+ SSL_CTX_set0_CA_list(ctx, SSL_load_client_CA_file(CAfile));
+
```
Let's show the preferred client-specific functions in the example.

mattcaswell commented:
@vdukhovni - new commit pushed addressing your feedback comments. Hopefully I covered them all.

vdukhovni commented on doc/man3/SSL_CTX_set0_CA_list.pod:
```diff
@@ -21,6 +39,69 @@ SSL_get0_peer_CA_list - get or set CA list
+
+The "generic CA list" functions below are very similar to the "client CA
+list" functions except that they have an effect on both the server and client
+sides. The lists of CA names managed are separate - so you cannot (for example)
```
Much better this time, thanks.  The only thing I'll ask is a short sentence here that repeats the admonition to avoid these functions unless CA names from client server are specifically required in the application protocol, in most cases only the above "client CA list." functions should be used.

mattcaswell replied:
Done - new commit pushed.

mattcaswell commented:
Pushed. Thanks.
